op-supervisor/supervisor/backend/db/iterator.go

```diff
@@ -0,0 +1,60 @@
+package db
+
+import (
+	"fmt"
+	"io"
+)
+
+type iterator struct {
+	db           *DB
+	nextEntryIdx int64
+
+	current logContext
+
+	entriesRead int64
+}
+
+func (i *iterator) NextLog() (blockNum uint64, logIdx uint32, evtHash TruncatedHash, outErr error) {
+	for i.nextEntryIdx <= i.db.lastEntryIdx() {
+		entryIdx := i.nextEntryIdx
+		entry, err := i.db.store.Read(entryIdx)
+		if err != nil {
+			outErr = fmt.Errorf("failed to read entry %v: %w", i, err)
+			return
+		}
+		i.nextEntryIdx++
+		i.entriesRead++
+		switch entry[0] {
+		case typeSearchCheckpoint:
+			current, err := newSearchCheckpointFromEntry(entry)
+			if err != nil {
+				outErr = fmt.Errorf("failed to parse search checkpoint at idx %v: %w", entryIdx, err)
+				return
+			}
+			i.current.blockNum = current.blockNum
+			i.current.logIdx = current.logIdx
+		case typeCanonicalHash:
+			// Skip
+		case typeInitiatingEvent:
+			evt, err := newInitiatingEventFromEntry(entry)
+			if err != nil {
+				outErr = fmt.Errorf("failed to parse initiating event at idx %v: %w", entryIdx, err)
```

Comment:
worth wrapping this in a custom error type. So it can easily be disambiguated from other io errors for data corruption handling

Reply:
The error returned from `newInitiatingeventFromEntry` will be a `ErrDataCorruption`  which I think is the right base type for this error - it should only happen if the database is corrupted.  Although in this case it can't happen because the only error that can be returned is that it wasn't a `typeInitiatingEvent but we already checked it was.